### PR TITLE
`SecurityContextDeny` admission plugin: add warning on creation and disable feature by default via a feature gate

### PR DESCRIPTION
**Pull request by mtardy**
#### What type of PR is this?

/kind deprecation

#### What this PR does / why we need it:

This PR adds a warning when enabling the `SecurityContextDeny` admission plugin because it's going to be removed. See https://github.com/kubernetes/kubernetes/issues/111516.

#### Which issue(s) this PR fixes:

This is step two of https://github.com/kubernetes/kubernetes/issues/111516.

#### Special notes for your reviewer:

~~I would need help on how to use [the warning system](https://github.com/kubernetes/enhancements/blob/master/keps/sig-api-machinery/1693-warnings/README.md) in order to issue a message when a Pod is created in a cluster with this admission plugin activated. See this message for more context:~~

> > Add a warning at runtime if the admission plugin is used with a reference to the docs of shortcomings
> 
> To me, I think the deprecation is strong enough that we could set a warning header on any Pod create that happens when the plugin is merely active (no matter whether the Pod was admitted or denied).
> 
> _Originally posted by @sftim in https://github.com/kubernetes/kubernetes/issues/111516#issuecomment-1408448738_

#### Does this PR introduce a user-facing change?

```release-note
The SecurityContextDeny admission plugin is going deprecated and will be removed in future versions.
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

```docs
- [KEP]: https://github.com/kubernetes/enhancements/issues/3785
- [Other doc]: https://k8s.io/docs/reference/access-authn-authz/admission-controllers/#securitycontextdeny
- [Issue]: https://github.com/kubernetes/kubernetes/issues/111516
```

### CODE REVIEW

**Comment by sftim** [comment]
Suggestion for changelog entry:
```markdown
Deprecated the alpha `SecurityContextDeny` admission plugin.

Added a warning on Pod creation whenever the deprecated `SecurityContextDeny` admission plugin
was active at the time the Pod was created.
```

**Comment by mtardy** [comment]
> Suggestion for changelog entry:
> 
> ```gfm
> Deprecated the alpha `SecurityContextDeny` admission plugin.
> 
> Added a warning on Pod creation whenever the deprecated `SecurityContextDeny` admission plugin
> was active at the time the Pod was created.
> ```

Thanks! But do you know actually how to achieve the second item? Because in the state, this PR only generates a warning on the activation of the plugin!

**Comment by sftim** [comment]
> do you know actually how to achieve the second item? Because in the state, this PR only generates a warning on the activation of the plugin!

https://github.com/kubernetes/kubernetes/blob/d6fe718e19881399fbb65f693c0c0f29f8f63861/plugin/pkg/admission/security/podsecurity/admission.go#L185 looks relevant. I don't code in Go very often and I've definitely never worked with this bit of Kubernetes code before.

File: plugin/pkg/admission/securitycontext/scdeny/admission.go [hunk 1]
```diff
@@ -45,6 +46,11 @@ var _ admission.ValidationInterface = &Plugin{}
 
 // NewSecurityContextDeny creates a new instance of the SecurityContextDeny admission controller
 func NewSecurityContextDeny() *Plugin {
+	// DEPRECATED: SecurityContextDeny will be removed in favor of PodSecurity admission.
+	klog.Warningf("%s admission controller is deprecated. "+
+		"Please remove this controller from your configuration files and scripts. "+
+		"See https://k8s.io/docs/reference/access-authn-authz/admission-controllers/#securitycontextdeny for more information.",
```
Comment (mikedanese) [inline plugin/pkg/admission/securitycontext/scdeny/admission.go:52]: We should update https://kubernetes.io/docs/reference/access-authn-authz/admission-controllers/#securitycontextdeny to explicitly say that the admission plugin is deprecated.

Comment (mtardy) [reply plugin/pkg/admission/securitycontext/scdeny/admission.go:52]: Ok I already updated that doc here https://github.com/kubernetes/website/pull/39131 😅 but you mention a big "deprecation" warning I guess.

**Comment by mikedanese** [comment]
We should add an alpha feature gate for 1.27 to guard SCDeny admission controller behind, then remove in 1.28. That + logging should be a good enough off ramp for this deprecation.

**Comment by mtardy** [comment]
> @mtardy: The following test **failed**, say `/retest` to rerun all failed tests or `/retest-required` to rerun all mandatory failed tests:
> 
> Test name	Commit	Details	Required	Rerun command
> pull-kubernetes-integration	[72e37a5](https://github.com/kubernetes/kubernetes/commit/72e37a5bb9a5c02db2199704a88172d7be72cbec)	[link](https://prow.k8s.io/view/gs/kubernetes-jenkins/pr-logs/pull/115879/pull-kubernetes-integration/1634238838748483584)	true	`/test pull-kubernetes-integration`
> [Full PR test history](https://prow.k8s.io/pr-history?org=kubernetes&repo=kubernetes&pr=115879). [Your PR dashboard](https://prow.k8s.io/pr?query=is%3Apr%20state%3Aopen%20author%3Amtardy). Please help us cut down on flakes by [linking to](https://git.k8s.io/community/contributors/devel/sig-testing/flaky-tests.md#filing-issues-for-flaky-tests) an [open issue](https://github.com/kubernetes/kubernetes/issues?q=is:issue+is:open) when you hit one in your PR.

ok so blocking registration is not a good idea...

File: plugin/pkg/admission/securitycontext/scdeny/admission.go [hunk 2]
```diff
@@ -23,17 +23,26 @@ import (
...
 	api "k8s.io/kubernetes/pkg/apis/core"
+	"k8s.io/kubernetes/pkg/features"
 )
 
 // PluginName indicates name of admission plugin.
 const PluginName = "SecurityContextDeny"
 
 // Register registers a plugin
 func Register(plugins *admission.Plugins) {
-	plugins.Register(PluginName, func(config io.Reader) (admission.Interface, error) {
```
Comment (liggitt) [inline plugin/pkg/admission/securitycontext/scdeny/admission.go:34]: check the feature gate inside the constructor and error with a message about this being an alpha feature, requiring the SecurityContextDeny feature gate to be enabled, and that the feature is planned to be removed:

```go
plugins.Register(PluginName, func(config io.Reader) (admission.Interface, error) {
  if utilfeature.DefaultFeatureGate.Enabled(features.SecurityContextDeny) {
    return nil, fmt.Errorf("...")
  }
  return NewSecurityContextDeny(), nil
})
```

File: plugin/pkg/admission/securitycontext/scdeny/admission.go [hunk 3]
```diff
@@ -23,17 +23,26 @@ import (
...
-		return NewSecurityContextDeny(), nil
-	})
+	if utilfeature.DefaultFeatureGate.Enabled(features.SecurityContextDeny) {
```
Comment (liggitt) [inline plugin/pkg/admission/securitycontext/scdeny/admission.go:37]: the feature gate isn't populated at the point in time Register is called... you need to check the gate inside the constructor

**Comment by mtardy** [comment]
I think it's ready @liggitt, I'll do another website PR to explicitly say that the thing is deprecated and we will be done for this small thing for v1.27 :)

**Comment by mtardy** [comment]
The Changelog entry could be similar to this:
```
Deprecated the alpha `SecurityContextDeny` admission plugin.

Added a warning on API server startup whenever the deprecated `SecurityContextDeny` admission plugin
is enabled. Added the `SecurityContextDeny` feature gate, disabled by default, to allow registering the plugin.
```

**Comment by sftim** [comment]
My idea:
```diff
-Deprecated the alpha `SecurityContextDeny` admission plugin.
+Deprecated the `SecurityContextDeny` admission plugin, which was previously alpha.
 
 Added a warning on API server startup whenever the deprecated `SecurityContextDeny` admission plugin
-is enabled. Added the `SecurityContextDeny` feature gate, disabled by default, to allow registering the plugin.
+is enabled. Added the `SecurityContextDeny` feature gate, disabled by default, to allow loading the plugin.
+To use the deprecated plugin you must both enable the feature gate and explicitly list it as an API server
+admission plugin.
```

**Pull request merged**